### PR TITLE
reverse most AreaShape functions

### PR DESCRIPTION
**Pull request by tetraxile**

### CODE REVIEW

File: lib/al/Library/Area/AreaShape.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#include <math/seadMatrix.h>
+#include <math/seadVector.h>
+
+namespace al {
+
+class AreaShape : public al::HioNode {
+private:
+    const sead::Matrix34f* mBaseMtxPtr = nullptr;
+
+protected:
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShape.h]: `private` instead, using a `getScale() const`

File: lib/al/Library/Area/AreaShape.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+bool AreaShape::calcWorldPos(sead::Vector3f* worldPos, const sead::Vector3f& trans) const {
+    if (al::isNearZero(mScale.x, 0.001f)) return false;
+    if (al::isNearZero(mScale.y, 0.001f)) return false;
+    if (al::isNearZero(mScale.z, 0.001f)) return false;
+
+    worldPos->x = trans.x * mScale.x;
+    worldPos->y = trans.y * mScale.y;
+    worldPos->z = trans.z * mScale.z;
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShape.cpp:55]:
```suggestion
    worldPos = trans * mScale;
```
Might give a mismatch, please check

File: lib/al/Library/Area/AreaShape.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+bool AreaShape::calcWorldDir(sead::Vector3f* worldDir, const sead::Vector3f& trans) const {
+    if (al::isNearZero(mScale.x, 0.001f)) return false;
+    if (al::isNearZero(mScale.y, 0.001f)) return false;
+    if (al::isNearZero(mScale.z, 0.001f)) return false;
+
+    worldDir->x = trans.x * mScale.x;
+    worldDir->y = trans.y * mScale.y;
+    worldDir->z = trans.z * mScale.z;
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShape.cpp:73]: same as above

File: lib/al/Library/Area/AreaShapeCube.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include "al/Library/Area/AreaShape.h"
+
+namespace al {
+
+class AreaShapeCube : public al::AreaShape {
+protected:
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShapeCube.h]: `private`

File: lib/al/Library/Area/AreaShapeCube.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include "al/Library/Area/AreaShape.h"
+
+namespace al {
+
+class AreaShapeCube : public al::AreaShape {
+protected: 
+    enum class OriginType { Center, Base, Top };
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShapeCube.h:9]: `public`

File: lib/al/Library/Area/AreaShapeCube.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    AreaShapeCube::OriginType mOriginType;
+
+    AreaShapeCube(al::AreaShapeCube::OriginType);
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShapeCube.h:15]: `public`

File: lib/al/Library/Area/AreaShapeCylinder.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace al {
+
+class AreaShapeCylinder : public al::AreaShape {
+protected: 
+    enum class OriginType { Center, Base, Top };
+
+    AreaShapeCylinder::OriginType mOriginType;
+
+    AreaShapeCylinder(al::AreaShapeCylinder::OriginType);
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShapeCylinder.h]: same as above

File: lib/al/Library/Area/AreaShapeOval.cpp [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace al {
+AreaShapeOval::AreaShapeOval() {}
+
+bool AreaShapeOval::isInVolume(const sead::Vector3f& trans) const {
+    sead::Vector3f localPos = sead::Vector3f::zero;
+    calcLocalPos(&localPos, trans);
+    float radius = 500.0f;
+
+    return localPos.squaredLength() <= radius * radius;
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShapeOval.cpp]:
```suggestion
    return localPos.squaredLength() <= sead::Mathf::square(500);
```

File: lib/al/Library/Area/AreaShapeSphere.cpp [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+namespace al {
+AreaShapeSphere::AreaShapeSphere() {}
+
+bool AreaShapeSphere::isInVolume(const sead::Vector3f& trans) const {
+    sead::Vector3f baseTrans;
+    calcTrans(&baseTrans);
+    sead::Vector3f offsetTrans = trans - baseTrans;
+    f32 radius = mScale.x * 500.0f;
+
+    return offsetTrans.squaredLength() <= radius * radius;
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShapeSphere.cpp]:
```suggestion
    return offsetTrans.squaredLength() <= sead::Mathf::square(getScale().x * 500);
```

File: lib/al/Library/Area/AreaShapeSphere.cpp [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+bool AreaShapeSphere::isInVolumeOffset(const sead::Vector3f& trans, f32 offset) const {
+    sead::Vector3f baseTrans;
+    calcTrans(&baseTrans);
+    sead::Vector3f offsetTrans = trans - baseTrans;
+    f32 radius = mScale.x * 500.0f + offset;
+
+    return offsetTrans.squaredLength() <= radius * radius;
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShapeSphere.cpp]:
```suggestion
    return offsetTrans.squaredLength() <= sead::Mathf::square(getScale().x * 500);
```

File: lib/al/Library/LiveActor/ActorFactory.h [hunk 11]
```diff
@@ -5,7 +5,9 @@
 namespace al {
 class LiveActor;
 
-class ActorFactory : public al::Factory<CreatorFunction<LiveActor>> {
+typedef LiveActor* (*ActorCreatorFunction)(const char* actorName);
```
Comment (MonsterDruide1) [inline lib/al/Library/LiveActor/ActorFactory.h]: `using` instead

File: lib/al/Library/Matrix/MatrixUtil.h [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+struct MatrixColumnMajor4x4fType;
+}
+
+namespace al {
+
+template <typename T>
+struct Matrix43 {
+    T m[3][4];
+};
+typedef Matrix43<f32> Matrix43f;
```
Comment (MonsterDruide1) [inline lib/al/Library/Matrix/MatrixUtil.h]: `using` instead

**Review by MonsterDruide1 — CHANGES_REQUESTED** [review]

File: lib/al/Library/Area/AreaShape.cpp (hunk 2 above)
Comment (tetraxile) [reply lib/al/Library/Area/AreaShape.cpp:55]: there doesn't currently exist an `operator*` for multiplying two `sead::Vector3`s; there are only those for dot products and matrix vector multiplication

File: lib/al/Library/Area/AreaShapeSphere.cpp (hunk 9 above)
Comment (tetraxile) [reply lib/al/Library/Area/AreaShapeSphere.cpp]: this causes a mismatch, seen here: https://decomp.me/scratch/HoRXn

File: lib/al/Library/Area/AreaShapeSphere.cpp (hunk 10 above)
Comment (tetraxile) [reply lib/al/Library/Area/AreaShapeSphere.cpp]: same as above (with `+ offset` added)

File: lib/al/Library/Area/AreaShapeSphere.cpp (hunk 9 above)
Comment (MonsterDruide1) [reply lib/al/Library/Area/AreaShapeSphere.cpp]: Alright, then you should to with the two-lined variant, but still use `square` instead:
https://decomp.me/scratch/0HdGu

File: lib/al/Library/Area/AreaShape.cpp (hunk 2 above)
Comment (MonsterDruide1) [reply lib/al/Library/Area/AreaShape.cpp:55]: Ehh, good point. Ignore this suggestion then.

File: lib/al/Library/Area/AreaShape.cpp (hunk 3 above)
Comment (MonsterDruide1) [reply lib/al/Library/Area/AreaShape.cpp:73]: (same here, ignore)

File: lib/al/Library/Area/AreaShapeCube.cpp [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+// bool AreaShapeCube::isInVolumeOffset(const sead::Vector3f& trans, f32 offset) const {}
+// bool AreaShapeCube::calcNearestEdgePoint(sead::Vector3f* edgePoint, const sead::Vector3f& trans) const {}
+// bool AreaShapeCube::checkArrowCollision(sead::Vector3f*, sead::Vector3f*, const sead::Vector3f&, const sead::Vector3f&) const {}
+// bool AreaShapeCube::calcLocalBoundingBox(sead::BoundBox3f* boundingBox) const {}
+
+bool AreaShapeCube::isInLocalVolume(const sead::Vector3f& trans) const {
+    float bottom = mOriginType == OriginType::Base ? 0.0f : (mOriginType == OriginType::Top ? -1000.0f : 500.0f);
+    float top = mOriginType == OriginType::Base ? 1000.0f : (mOriginType == OriginType::Top ? 0.0f : 500.0f);
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShapeCube.cpp:21]: Doing a `switch` might be easier to look at, might even combine `bottom` and `top` in a single `switch`.

File: lib/al/Library/Area/AreaShapeCube.cpp [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    sead::Vector3f min = { -500.0f, bottom, -500.0f };
+    sead::Vector3f max = { 500.0f, top, 500.0f };
+
+    if (
+        (trans.y < min.y || max.y < trans.y) ||
+        (trans.x < min.x || max.x < trans.x) ||
+        (trans.z < min.z || max.z < trans.z)
+    ) return false;
+
+    return true;
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShapeCube.cpp]: Create a `sead::BoundBox3f` and call `isInside()`, might match too

**Review by MonsterDruide1 — CHANGES_REQUESTED** [review]

File: lib/al/Library/Area/AreaShapeCube.cpp (hunk 14 above)
Comment (tetraxile) [reply lib/al/Library/Area/AreaShapeCube.cpp]: i tried this and can't get it to match no matter how i do it unfortunately

Comment (MonsterDruide1) [reply lib/al/Library/Area/AreaShapeCube.cpp]: Can you create a `decomp.me` of it?

File: lib/al/Library/Area/AreaShapeCube.cpp (hunk 13 above)
Comment (tetraxile) [reply lib/al/Library/Area/AreaShapeCube.cpp:21]: `if`/`else if`/`else` works here, i can't get it to match when combining them into one but it works when they're separate. `switch` works too but i think `if`-`else` looks better personally. i'll combine it with the suggestion below on decomp.me

File: lib/al/Library/Area/AreaShapeCube.cpp (hunk 14 above)
Comment (tetraxile) [reply lib/al/Library/Area/AreaShapeCube.cpp]: https://decomp.me/scratch/dG4Vo

**Review by MonsterDruide1 — APPROVED** [review]

File: lib/al/Library/Area/AreaShape.cpp [hunk 15]
```diff
@@ -0,0 +1,98 @@
+#include "al/Library/Area/AreaShape.h"
+#include "al/Library/Area/AreaShapeCube.h"
+#include "al/Library/Area/AreaShapeCylinder.h"
+#include "al/Library/Area/AreaShapeInfinite.h"
+#include "al/Library/Area/AreaShapeSphere.h"
+#include "al/Library/Math/MathAngleUtil.h"
+#include "al/Library/Math/MathLengthUtil.h"
+#include "al/Library/Matrix/MatrixUtil.h"
+#include <math/seadVector.hpp>
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShape.cpp]:
```suggestion
#include <math/seadVector.h>
```
clang-format to re-order includes

File: lib/al/Library/Area/AreaShape.h [hunk 16]
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include "al/Library/Factory/Factory.h"
+#include "al/Library/HostIO/HioNode.h"
+#include <basis/seadTypes.h>
+#include <math/seadBoundBox.h>
+#include <math/seadMatrix.h>
+#include <math/seadVector.h>
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShape.h]: clang-format to re-order includes

File: lib/al/Library/Area/AreaShape.h [hunk 17]
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include "al/Library/Factory/Factory.h"
+#include "al/Library/HostIO/HioNode.h"
+#include <basis/seadTypes.h>
+#include <math/seadBoundBox.h>
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShape.h:3]: useless include

File: lib/al/Library/Area/AreaShape.h [hunk 18]
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include "al/Library/Factory/Factory.h"
+#include "al/Library/HostIO/HioNode.h"
+#include <basis/seadTypes.h>
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShape.h]: useless include

File: lib/al/Library/Area/AreaShapeCube.cpp [hunk 19]
```diff
@@ -0,0 +1,36 @@
+#include "al/Library/Area/AreaShapeCube.h"
+#include <math/seadBoundBox.h>
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShapeCube.cpp]: useless include

File: lib/al/Library/Area/AreaShapeOval.cpp [hunk 20]
```diff
@@ -0,0 +1,35 @@
+#include "al/Library/Area/AreaShapeOval.h"
+#include "al/Library/Area/AreaShapeSphere.h"
```
Comment (MonsterDruide1) [inline lib/al/Library/Area/AreaShapeOval.cpp]: useless include

File: lib/al/Library/Math/MathAngleUtil.h [hunk 21]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <basis/seadTypes.h>
```
Comment (MonsterDruide1) [inline lib/al/Library/Math/MathAngleUtil.h]: useless include

File: lib/al/Library/Math/MathLengthUtil.h [hunk 22]
```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include <basis/seadTypes.h>
```
Comment (MonsterDruide1) [inline lib/al/Library/Math/MathLengthUtil.h]: useless include

File: lib/al/Library/Matrix/MatrixPtrHolder.h [hunk 23]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <basis/seadTypes.h>
```
Comment (MonsterDruide1) [inline lib/al/Library/Matrix/MatrixPtrHolder.h]: useless include

File: lib/al/Library/Matrix/MatrixUtil.h [hunk 24]
```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include <basis/seadTypes.h>
```
Comment (MonsterDruide1) [inline lib/al/Library/Matrix/MatrixUtil.h]: useless include

**Review by MonsterDruide1 — CHANGES_REQUESTED** [review]

**Review by MonsterDruide1 — APPROVED** [review]

**Pull request merged**